samples/WinForms/Dx12Viewport.cs

```diff
@@ -0,0 +1,769 @@
+{
+    public unsafe class DX12Viewport : IDisposable
+    {
+
+        // Viewport dimensions
+        private uint _width;
+
+        private uint _height;
+
+        private float _aspectRatio;
+
+        // Adapter info
+        private readonly bool _useWarpDevice;
+
+        // Root assets path
+        private readonly string _assetsPath;
+
+        private const uint FrameCount = 2;
+
+        // Pipeline objects
+        private D3D12_VIEWPORT _viewport;
+        private RECT _scissorRect;
+        private IDXGISwapChain3* _swapChain;
+        private ID3D12Device* _device;
+        private readonly ID3D12Resource*[] _renderTargets;
+        private ID3D12CommandAllocator* _commandAllocator;
+        private ID3D12CommandQueue* _commandQueue;
+        private ID3D12RootSignature* _rootSignature;
+        private ID3D12DescriptorHeap* _rtvHeap;
+        private ID3D12PipelineState* _pipelineState;
+        private ID3D12GraphicsCommandList* _commandList;
+        private uint _rtvDescriptorSize;
+        private readonly IntPtr _hWnd;
+        private bool _isOnColorToggle;
+
+        // App resources.
+        private ID3D12Resource* _vertexBuffer;
+        private D3D12_VERTEX_BUFFER_VIEW _vertexBufferView;
+
+        // Synchronization objects.
+        private uint _frameIndex;
+        private HANDLE _fenceEvent;
+        private ID3D12Fence* _fence;
+        private ulong _fenceValue;
+
+
+        public uint Width => _width;
+
+        public uint Height => _height;
+
+        public float AspectRatio => _aspectRatio;
+
+        public bool UseWarpDevice => _useWarpDevice;
+
+        public DX12Viewport(uint width, uint height, IntPtr hWnd, bool useWarpDevice = false)
+        {
+            _width = width;
+            _height = height;
+            _assetsPath = GetAssetsPath();
+            _aspectRatio = width / ((float)height);
+
+            _renderTargets = new ID3D12Resource*[FrameCount];
+
+            _viewport = new D3D12_VIEWPORT {
+                TopLeftX = 0,
+                TopLeftY = 0,
+                Width = width,
+                Height = height,
+                MinDepth = D3D12_MIN_DEPTH,
+                MaxDepth = D3D12_MAX_DEPTH
+            };
+
+            _scissorRect = new RECT {
+                left = 0,
+                top = 0,
+                right = unchecked((int)width),
+                bottom = unchecked((int)height)
+            };
+
+            _hWnd = hWnd;
+            _useWarpDevice = useWarpDevice;
+            _isOnColorToggle = false;
+        }
+
+
+        public virtual void OnResize(uint width, uint height)
+        {
+            _width = width;
+            _height = height;
+            _aspectRatio = width / ((float)height);
+        }
+
+        // Samples override the event handlers to handle specific messages
+        public virtual void OnKeyDown(byte key)
+        {
+        }
+
+        public virtual void OnKeyUp(byte key)
+        {
+        }
+
```

Comment:
There is a newline here that needs to be removed
```suggestion
```

Reply:
ok